library/core/src/stream/stream/mod.rs

```diff
@@ -0,0 +1,129 @@
+mod next;
+
+pub use next::Next;
+
+use crate::ops::DerefMut;
+use crate::pin::Pin;
+use crate::task::{Context, Poll};
+
+/// An interface for dealing with asynchronous iterators.
+///
+/// This is the main stream trait. For more about the concept of streams
+/// generally, please see the [module-level documentation]. In particular, you
+/// may want to know how to [implement `Stream`][impl].
+///
+/// [module-level documentation]: index.html
+/// [impl]: index.html#implementing-stream
+#[unstable(feature = "async_stream", issue = "79024")]
+#[must_use = "streams do nothing unless polled"]
+pub trait Stream {
+    /// The type of items yielded by the stream.
+    type Item;
+
+    /// Attempt to pull out the next value of this stream, registering the
+    /// current task for wakeup if the value is not yet available, and returning
+    /// `None` if the stream is exhausted.
+    ///
+    /// # Return value
+    ///
+    /// There are several possible return values, each indicating a distinct
+    /// stream state:
+    ///
+    /// - `Poll::Pending` means that this stream's next value is not ready
+    /// yet. Implementations will ensure that the current task will be notified
+    /// when the next value may be ready.
+    ///
+    /// - `Poll::Ready(Some(val))` means that the stream has successfully
+    /// produced a value, `val`, and may produce further values on subsequent
+    /// `poll_next` calls.
+    ///
+    /// - `Poll::Ready(None)` means that the stream has terminated, and
+    /// `poll_next` should not be invoked again.
```

Comment:
Given that iterators [may choose to resume iteration](https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#tymethod.next), I think maybe we should relax this constraint?

Reply:
Calling `poll_next` after `Poll::Ready(None)` is returned can result in a panic, as described in the "panics" section. (This is especially true when [implementing streams based on generators](https://github.com/taiki-e/futures-async-stream/blob/f516f3a2a77182e99ece456f6e3a5ab0c18c115a/src/lib.rs#L346-L362).)
I think it's okay to create a stream implementation that doesn't end with `Poll::Ready(None)`, but it is preferable to assume all streams end with `Poll::Ready(None)`, except in the rare case where you know the concrete type and it is guaranteed that stream doesn't end with `Poll::Ready(None)`.

Reply:
...So I think relaxing this constraint would have little benefit to the user and would only confuse it.
(Is there a real use case that benefits from the relaxation of this constraint?)

Reply:
(I personally believe that this is exactly the same thing as specified in `Iterator::next` docs, there is no constraint on what an iterator does once it is "finished" so it is perfectly valid for it to panic if polled again; this is just much more explicit about it.)

Reply:
Hmm, if what @Nemo157 is true then I think we should update the Iterator docs to be more explicit. They don't rule out that an iterator may panic, but also don't establish it as a normal thing you should assume in the general case.

@taiki-e My worry was that people would assume that the Iterator contract around this would carry over to Stream (which is a very reasonable assumption IMO). But now it's unclear to me what the contract actually is.

Reply:
I find it a bit odd the contention that its valid for `Iterator::next` to panic, but also return `Some` after it has returned `None`,
given that [FusedIterator](https://doc.rust-lang.org/std/iter/trait.FusedIterator.html)
> Calling next on a fused iterator that has returned None once is guaranteed to return None again.

And anywhere you panic you *could* just as easily return `None`.
There are cases where I have used Non-fused iterators which return `Some`, after returning `None`, It is weird that in the above interpretation panic'ing would be allowed for a non-fused iterator, but not for a fused one, since it is guaranteed to return `None`.

Reply:
@ratmice FusedIterator is a separate trait from Iterator, so both are true at once: you can have an iterator that returns Some, None, Some, it just won't be a FusedIterator.

I agree though that it's better to return None than to panic - I guess the issue here is there's no equivalent of None for the Stream to return? Can't it just return `Stream::Ready(None)`?

Reply:
@jyn514 Yeah, my point was largely that if you `panic!` on next instead of implementing `FusedIterator`, the `Some`, `None`, `Some` allowance on `Iterator` becomes fairly broken, in these cases it *should* implement `FusedIterator` which does preclude `panic!`.

I have an implementation of a Non-Fused DelayedIterator which implements None while count > 0, and only then starts iterating.  It's largely useless given `slice::windows` not sure if something like that would ever be useful for stream?

I guess my question largely is whether we should try and keep the `Fused`/NonFused distinction intact.  As it seems to me this `stream` is behaving more like `Fused` with a panic instead of repeating None.

Reply:
Part of @taiki-e's point was that generators panic if you resume them after they return. So adapting from generator to Stream would have to involve storing another bit of information (whether it has returned) if panics aren't allowed.

But the generator objects already have that information (it's how they know to panic), and we could find some way to expose it to the Stream adaptor code, if we wanted.

Reply:
> @jyn514 Yeah, my point was largely that if you `panic!` on next instead of implementing `FusedIterator`, the `Some`, `None`, `Some` allowance on `Iterator` becomes fairly broken, in these cases it _should_ implement `FusedIterator` which does preclude `panic!`.

As I see it there are three major allowed `Iterator` and `Stream` implementations/requirements:

 1. A `FusedIterator` (and probably `FusedStream` in the future, though [`futures::stream::FusedStream`](https://docs.rs/futures/0.3.8/futures/stream/trait.FusedStream.html) works off a different system where it pre-checks whether it should be polled at all) which guarantees any amount of `Some` followed by an infinite number of `None`.
 2. A "resumeable" iterator which guarantees a mix of `Some` and `None`, and documents under what conditions it switches between them.
 3. A "single-use" iterator which guarantees any amount of `Some`, followed by a single `None`, and then doesn't guarantee its behaviour afterwards.

Distinguishing between the second and third variants is only a matter of documentation, and the majority of the time all that's needed is the third variant, resuming an iterator seems to be a very uncommon thing, especially in a context where you have a generic `I: Iterator` since you don't know the exact conditions under which it switches.

> Part of @taiki-e's point was that generators panic if you resume them after they return. So adapting from generator to Stream would have to involve storing another bit of information (whether it has returned) if panics aren't allowed.

And this will be relevant for iterators in the future when we have generator-backed `Iterator` implementations.

Reply:
It sounds like we have consensus that `Iterator` and `Stream` are allowed to return items after they've yielded `None` once. But in practice they rarely do, and when created through generators they never will.

I'm inclined to omit the `panic` section for now, and make sure the text from `Stream::next` matches the text from `Iterator::next`. Then in a follow-up PR add a `# panic` heading to both which clarifies that these methods may panic. We can add this as a concern to the tracking issue to make sure this is clarified and consistent between `Iterator` and `Stream` before stabilization.

Does that sound like an acceptable outcome?

Reply:
> I'm inclined to omit the `panic` section for now, and make sure the text from `Stream::next` matches the text from `Iterator::next`. Then in a follow-up PR add a `# panic` heading to both which clarifies that these methods may panic. We can add this as a concern to the tracking issue to make sure this is clarified and consistent between `Iterator` and `Stream` before stabilization.

I'm in favor of adding a panic section to Iterator, but I don't think it's necessary to remove the panic section from Stream in this patch. The current panic section in this patch is consistent with `Future` and explains exactly what can happen.

Reply:
Ensuring the documentation of panic behavior on `Stream` and `Iterator` is consistent should be something we resolve before stabilization. I've added an entry for this to the tracking issue: https://github.com/rust-lang/rust/issues/79024.

Going to mark this thread as resolved as this is something which can be addressed in follow-up PRs (and likely requires updates to the `Iterator` docs too).